src/script/interpreter.cpp

```diff
@@ -944,20 +1015,20 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
                             fSuccess = false;
                     }
 
+                    // The only way in which a CHECKMULTISIG can fail without aborting
+                    // execution is by leaving all signatures empty (and simultaneously
+                    // marking all pubkeys as skipped, but that is already checked for).
+                    if (require_valid_sigs && !fSuccess && !empty_sigs) {
+                        return set_error(serror, SCRIPT_ERR_FAILED_SIGNATURE_CHECK);
+                    }
+
                     // Clean up stack of actual arguments
                     while (i-- > 1)
                         popstack(stack);
 
                     // A bug causes CHECKMULTISIG to consume one extra argument
-                    // whose contents were not checked in any way.
-                    //
-                    // Unfortunately this is a potential source of mutability,
-                    // so optionally verify it is exactly equal to zero prior
-                    // to removing it from the stack.
                     if (stack.size() < 1)
                         return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
-                    if ((flags & SCRIPT_VERIFY_NULLDUMMY) && stacktop(-1).size())
-                        return set_error(serror, SCRIPT_ERR_SIG_NULLDUMMY);
```

Comment:
Is it a bad idea to allow both nulldummy and the bitfield style? They seem compatible with each other. Allowing nulldummy would mean greater compat with Bitcoin.

Reply:
Well obviously they can never both be enabled at the same time. But do you mean just leave the dead code that uses `SCRIPT_VERIFY_NULLDUMMY`? I'll look into that. Note that it is safer to remove if we expect that this will never make it into bitcoin (which already enforces `SCRIPT_VERIFY_NULLDUMMY` in witness scripts, but that could be handled differently in future script updates).

Reply:
Actually, for some reason I assumed that people could choose one or the other (e.g. a segwit compat wallet could use the segwit style of nulldummy policy), but I realize this is per node, not per tx, so scratch that idea.

Reply:
Seems relatively straight forward to keep both. I'll do that. It has the advantage of not having to update the old tests, but rather just create new tests, which makes it easier to maintain.

Reply:
Not per node, but per script validation mode. In traditional and P2SH scripts `NULLDUMMY` is not enforced. In bitcoin's segwit v0 it is enforced. But bitcoin could switch to `REQUIRE_VALID_SIGS` in v1 witness scripts, for example.

Reply:
Ohh, right, that's the way it works. Thanks.